File: drive/src/main/java/de/mein/drive/bash/BashToolsUnix.java
package de.mein.drive.bash;

import de.mein.Lok;
import de.mein.auth.file.AFile;
import de.mein.auth.tools.N;
import de.mein.auth.file.DefaultFileConfiguration;

import org.jdeferred.Promise;
import org.jdeferred.impl.DeferredObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * Created by xor on 13.07.2017.
 */
public class BashToolsUnix implements BashToolsImpl {

    protected String BIN_PATH = "/bin/sh";
    private ExecutorService executorService = Executors.newCachedThreadPool();


    public static void main(String[] args) throws Exception {
        AFile.configure(new DefaultFileConfiguration());
        AFile f = AFile.instance("f");
        f.mkdirs();
        BashToolsUnix bashToolsUnix = new BashToolsUnix();
        ModifiedAndInode modifiedAndInode = bashToolsUnix.getModifiedAndInode(f);
        Lok.debug("mod " + modifiedAndInode.getModified() + " ... inode " + modifiedAndInode.getiNode());
        f.delete();
    }


    public ModifiedAndInode getModifiedAndInode(AFile file) throws IOException {
        String[] args = new String[]{BIN_PATH, "-c", "stat -c %Y\" \"%i " + escapeAbsoluteFilePath(file)};
        Process proc = new ProcessBuilder(args).start();
        BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
        String line = reader.readLine();
        String[] split = line.split(" ");
        return new ModifiedAndInode(Long.parseLong(split[0]), Long.parseLong(split[1]));
    }


    @Override
    public void setBinPath(String binPath) {
        BIN_PATH = binPath;
    }

    @Override
    public Set<Long> getINodesOfDirectory(AFile file) throws IOException {
        String[] args = new String[]{BIN_PATH, "-c", "find share/ -printf \"%p\\n\" | tail -n +2 | xargs -d '\\n' stat -c %i"};
        Process proc = new ProcessBuilder(args).start();
        BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
        Set<Long> iNodes = new HashSet<>();
        String line = "";
        while ((line = reader.readLine()) != null) {
            iNodes.add(Long.parseLong(line));
        }
        return iNodes;
    }

    /**
     * escapes this character: "
     *
     * @param file
     * @return
     */
    protected String escapeAbsoluteFilePath(AFile file) {
        return "\"" + file.getAbsolutePath()
                .replaceAll("\"", "\\\\\"")
                .replaceAll("`", "\\\\`")
                .replaceAll("\\$", "\\\\\\$")
                + "\"";
    }

    @Override
    public ModifiedAndInode getModifiedAndINodeOfFile(AFile file) throws IOException, InterruptedException {
        String[] args = new String[]{BIN_PATH, "-c", "stat -c %i\\ %Y " + escapeAbsoluteFilePath(file)};
        Process proc = new ProcessBuilder(args).start();
        //proc.waitFor(); // this line sometimes hangs. Process.exitcode is 0 and Process.hasExited is false
        BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
        String line = reader.readLine();
        //todo debug
        if (line == null) {
            Lok.debug("reading error for: " + args[2]);
            try {
                BufferedReader r = new BufferedReader((new InputStreamReader(proc.getErrorStream())));
                String l = r.readLine();
                while (l != null) {
                    Lok.debug(l);
                    l = r.readLine();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        String[] parts = line.split(" ");
        Long iNode = Long.parseLong(parts[0]);
        Long modified = Long.parseLong(parts[1]);
        return new ModifiedAndInode(modified, iNode);
    }

    /**
     * rm -rf
     *
     * @param directory
     */
    @Override
    public void rmRf(AFile directory) throws IOException {
        String[] args = new String[]{BIN_PATH, "-c", "rm -rf " + escapeAbsoluteFilePath(directory)};
        Process proc = new ProcessBuilder(args).start();
        N.oneLine(proc::waitFor);
    }

    @Override
    public List<AFile> stuffModifiedAfter(AFile referenceFile, AFile directory, AFile pruneDir) throws IOException, BashToolsException {
        Lok.debug("BashTools.stuffModifiedAfter: " + referenceFile.getName() + " mod: " + referenceFile.lastModified());
//        String cmd = "find \"" + directory.getAbsolutePath() + "\"  "
//                + " -path \"" + pruneDir + "\" -prune"
//                + " -o -newer \"" + referenceFile.getAbsolutePath() + "\" -print";
        String cmd = "find " + escapeAbsoluteFilePath(directory)
                + " -path " + escapeAbsoluteFilePath(pruneDir) + " -prune"
                + " -o -newer " + escapeAbsoluteFilePath(referenceFile) + " -print";
        Lok.debug("BashTools.stuffModifiedAfter.cmd: " + cmd);
        String[] args = new String[]{BIN_PATH, "-c",
                cmd};
        ProcessBuilder processBuilder = new ProcessBuilder(args);
        processBuilder.redirectErrorStream(true);
        Process proc = processBuilder.start();
        Lok.debug("BashTools.stuffModifiedAfter.collecting.result");
        List<AFile> result = new ArrayList<>();
        Iterator<AFile> iterator = BashTools.inputStreamToFileIterator(proc.getInputStream());
        while (iterator.hasNext()) {
            AFile path = iterator.next();
            Lok.debug(getClass().getSimpleName() + ".stuffModifiedAfter.collected: " + path);
            result.add(path);
        }
        Lok.debug("BashTools.stuffModifiedAfter.collecting.done");
        return result;
    }

    private Iterator<AFile> exec(String cmd) throws IOException {
        String[] args = new String[]{BIN_PATH, "-c",
                cmd};
        Lok.debug("BashToolsUnix.exec: " + cmd);
        Process proc = new ProcessBuilder(args).start();
        return BashTools.inputStreamToFileIterator(proc.getInputStream());
    }

    @Override
    public Iterator<AFile> find(AFile directory, AFile pruneDir) throws IOException {
        return exec("find " + escapeAbsoluteFilePath(directory) + " -mindepth 1" + " -path " + escapeAbsoluteFilePath(pruneDir) + " -prune -o -print");
    }

    @Override
    public Promise<Long, Exception, Void> getInode(AFile f) {
        DeferredObject<Long, Exception, Void> deferred = new DeferredObject<>();
        executorService.execute(() -> N.r(() -> {
            String ba = "ls -i -d " + escapeAbsoluteFilePath(f);
            String[] args = new String[]{BIN_PATH, "-c", ba};
            Long inode;
            List<String> lines;
            boolean hasFinished = false;
            Process proc = null;
            while (!hasFinished) {
                try {
                    proc = new ProcessBuilder(args).start();
                    hasFinished = proc.waitFor(10, TimeUnit.SECONDS);
                    if (!hasFinished) {
                        BufferedReader errorReader = new BufferedReader(new InputStreamReader(proc.getErrorStream()));
                        List<String> errors = errorReader.lines().collect(Collectors.toList());
                        Lok.debug("BashTools.stuffModifiedAfter.did not finish");
                    }
                    // try to read anyway.
                    // the process might have come to an end but Process.waitFor() does not always finish.
                    Lok.debug("BashTools.stuffModifiedAfter");
                    BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
                    Lok.debug("BashTools.stuffModifiedAfter.collecting.result");
                    lines = reader.lines().collect(Collectors.toList());
                    lines.forEach(s -> Lok.debug("BashTools.getNodeAndTime.LLLL " + s));
                    String[] s = lines.get(0).split(" ");
                    if (s[0].isEmpty())
                        Lok.debug("BashTools.getNodeAndTime");
                    inode = Long.parseLong(s[0]);
                    deferred.resolve(inode);
                    hasFinished = true;
                } catch (Exception e) {
                    e.printStackTrace();
                    proc.destroyForcibly();
                    continue;
                }
            }
        }));
        return deferred;
    }

    @Override
    public Iterator<AFile> stuffModifiedAfter(AFile directory, AFile pruneDir, long timeStamp) {
        System.err.println("BashToolsUnix.stuffModifiedAfter()... I AM THE UNIX GUY! >:(");
        return null;
    }

    @Override
    public void mkdir(AFile dir) throws IOException {
        String[] args = new String[]{BIN_PATH, "-c",
                "mkdir " + escapeAbsoluteFilePath(dir)};
        new ProcessBuilder(args).start();
    }

    @Override
    public boolean mv(File source, File target) throws IOException {
        String src = source.getAbsolutePath().replaceAll("'", "\\'");
        String tgt = target.getAbsolutePath().replaceAll("'", "\\'");
        String cmd = "mv '" + src + "' '" + tgt + "'";
        String[] args = new String[]{BIN_PATH, "-c", cmd};
        Process process = new ProcessBuilder(args).start();
        try {
            process.waitFor();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        BufferedReader reader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
        String error = reader.readLine();
        if (error == null)
            return true;
        Lok.debug("BashToolsUnix.mv");
        return false;
    }


    public Long getInotifyLimit() throws IOException {
        // "cat /proc/sys/fs/inotify/max_user_watches";
        File f = new File("/proc/sys/fs/inotify/max_user_watches");
        List<String> lines = Files.readAllLines(Paths.get(f.toURI()));
        return Long.parseLong(lines.get(0));
//        String[] args = new String[]{BIN_PATH, "-c", "cat " + "/proc/sys/fs/inotify/max_user_watches"};
//        Process proc = new ProcessBuilder(args).start();
//        BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
//        String line = reader.readLine();
////        String[] split = line.split(" ");
//        Long limit = Long.parseLong(line);
//        return limit;
    }

    public Long countSubDirs(File dir) throws IOException {
        String[] args = new String[]{BIN_PATH, "-c", "find " + escapeAbsoluteFilePath(AFile.instance(dir)) + "  -mindepth 1  -type d | wc -l"};
        Process proc = new ProcessBuilder(args).start();
        BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
        String line = reader.readLine();
//        String[] split = line.split(" ");
        Long subDirCount = Long.parseLong(line);
        return subDirCount;
    }
}

File: drive/src/main/java/de/mein/drive/bash/BashToolsUnix.kt
package de.mein.drive.bash

import de.mein.Lok
import de.mein.auth.file.AFile
import de.mein.auth.tools.N
import de.mein.auth.file.DefaultFileConfiguration

import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import java.util.concurrent.Executors

/**
 * Created by xor on 13.07.2017.
 */
open class BashToolsUnix : BashToolsImpl {


    override fun isSymLink(f: AFile<out AFile<*>>?): Boolean {
        if (f == null)
            return false
        return Files.isSymbolicLink(Paths.get(f.absolutePath))
    }

    protected var BIN_PATH = "/bin/sh"
    private val executorService = Executors.newCachedThreadPool()


    val inotifyLimit: Long?
        @Throws(IOException::class)
        get() {
            val f = File("/proc/sys/fs/inotify/max_user_watches")
            val lines = Files.readAllLines(Paths.get(f.toURI()))
            return java.lang.Long.parseLong(lines[0])
        }


    override fun setBinPath(binPath: String) {
        BIN_PATH = binPath
    }

    @Throws(IOException::class)
    override fun getINodesOfDirectory(file: AFile<*>): Set<Long> {
        val args = arrayOf(BIN_PATH, "-c", "find share/ -printf \"%p\\n\" | tail -n +2 | xargs -d '\\n' stat -c %i")
        val proc = ProcessBuilder(*args).start()
        val reader = BufferedReader(InputStreamReader(proc.inputStream))
        val iNodes = HashSet<Long>()
        var line = reader.readLine()
        while (line != null) {
            iNodes.add(java.lang.Long.parseLong(line))
            line = reader.readLine()
        }
        return iNodes
    }

    /**
     * escapes this character: "
     *
     * @param file
     * @return
     */
    protected fun escapeQuotedAbsoluteFilePath(file: AFile<*>): String {
        return ("\"" + file.absolutePath
                .replace("\"".toRegex(), "\\\\\"")
                .replace("`".toRegex(), "\\\\`")
                .replace("\\$".toRegex(), "\\\\\\$")
                + "\"")
    }

    /**
     * escapes this character: "
     *
     * @param file
     * @return
     */
    protected fun escapeAbsoluteFilePath(file: AFile<*>): String {
        return file.absolutePath
                .replace("\"".toRegex(), "\\\\\"")
                .replace("`".toRegex(), "\\\\`")
                .replace("\\$".toRegex(), "\\\\\\$")
                .replace(" ".toRegex(), "\\ ")

    }

    @Throws(IOException::class, InterruptedException::class)
    override fun getFsBashDetails(file: AFile<*>): FsBashDetails {
        val args = arrayOf(BIN_PATH, "-c", "stat -c %i\\ %Y\\ '%F'\\ %N " + escapeQuotedAbsoluteFilePath(file))
        val proc = ProcessBuilder(*args).start()
        //proc.waitFor(); // this line sometimes hangs. Process.exitcode is 0 and Process.hasExited is false
        val reader = BufferedReader(InputStreamReader(proc.inputStream))
        val line = reader.readLine()
        //todo debug
        if (line == null) {
            Lok.debug("reading error for: " + args[2])
            try {
                val r = BufferedReader(InputStreamReader(proc.errorStream))
                var l: String? = r.readLine()
                while (l != null) {
                    Lok.debug(l)
                    l = r.readLine()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        val parts = line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val iNode = java.lang.Long.parseLong(parts[0])
        val modified = java.lang.Long.parseLong(parts[1])
        val symLink = parts[2].startsWith("sym")
        val name = file.name
        var symLinkTarget: String? = null
        if (symLink) {
            symLinkTarget = parts[6]
            symLinkTarget = symLinkTarget.drop(1).dropLast(1)
        }
        return FsBashDetails(modified, iNode, symLink, symLinkTarget, name)
    }

    /**
     * parses an output line from stat
     */
    private fun parseStatLine(line: String): FsBashDetails {
        val parts = line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val iNode = java.lang.Long.parseLong(parts[0])
        val modified = java.lang.Long.parseLong(parts[1])
        val symLink = parts[2].startsWith("sym")
        val name = "NAME"
        var symLinkTarget: String? = null
        if (symLink) {
            symLinkTarget = parts[6]
            symLinkTarget = symLinkTarget.drop(1).dropLast(1)
        }
        return FsBashDetails(modified, iNode, symLink, symLinkTarget, name)
    }

    override fun getContentFsBashDetails(directory: AFile<*>): MutableMap<String, FsBashDetails> {
        val map = mutableMapOf<String, FsBashDetails>()
        val path = "${escapeAbsoluteFilePath(directory)}${File.separator}*"
//        val args = arrayOf(BIN_PATH, "-c", "stat -c %i\\ %Y\\ '%F'\\ %N\\ %n " + path)
        val args = arrayOf(BIN_PATH, "-c", "stat -c %i\\ //\\ %Y\\ //\\ '%F'\\ //\\ %N\\ //\\ %n " + path)

        val proc = ProcessBuilder(*args).start()
        //proc.waitFor(); // this line sometimes hangs. Process.exitcode is 0 and Process.hasExited is false
        val reader = BufferedReader(InputStreamReader(proc.inputStream))
        proc.waitFor()
        if (proc.exitValue() == 0) {
            reader.lines().forEach { line ->
                val parts = line.split("\\ //\\ ".toRegex()).toTypedArray()

//                val parts = line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val iNode = java.lang.Long.parseLong(parts[0])
                val modified = java.lang.Long.parseLong(parts[1])
                val symLink = parts[2].startsWith("sym")
                val absolutePath = parts[4]
                val name = File(absolutePath).name
                var symLinkTarget: String? = null
                if (symLink) {
                    // parse something like: "'/a/b '-> '/c/d'"
                    // drop quotes, spaces etc
                    symLinkTarget = parts[3].drop(absolutePath.length + 7).dropLast(1)
                }
                val details = FsBashDetails(modified, iNode, symLink, symLinkTarget, name)
                map[details.name] = details
            }
        } else {
            args.forEach {
                Lok.error("arg: $it")
            }
            val r = BufferedReader(InputStreamReader(proc.errorStream))
            r.lines().forEach {
                Lok.error(it)
            }
        }
        return map
    }

    /**
     * rm -rf
     *
     * @param directory
     */
    @Throws(IOException::class)
    override fun rmRf(directory: AFile<*>) {
        val args = arrayOf(BIN_PATH, "-c", "rm -rf " + escapeQuotedAbsoluteFilePath(directory))
        val proc = ProcessBuilder(*args).start()
        N.oneLine { proc.waitFor() }
    }

    @Throws(IOException::class, BashToolsException::class)
    override fun stuffModifiedAfter(referenceFile: AFile<*>, directory: AFile<*>, pruneDir: AFile<*>): List<AFile<*>> {
        Lok.debug("BashTools.stuffModifiedAfter: " + referenceFile.name + " mod: " + referenceFile.lastModified())
        //        String cmd = "find \"" + directory.getAbsolutePath() + "\"  "
        //                + " -path \"" + pruneDir + "\" -prune"
        //                + " -o -newer \"" + referenceFile.getAbsolutePath() + "\" -print";
        val cmd = ("find " + escapeQuotedAbsoluteFilePath(directory)
                + " -path " + escapeQuotedAbsoluteFilePath(pruneDir) + " -prune"
                + " -o -newer " + escapeQuotedAbsoluteFilePath(referenceFile) + " -print")
        Lok.debug("BashTools.stuffModifiedAfter.cmd: $cmd")
        val args = arrayOf(BIN_PATH, "-c", cmd)
        val processBuilder = ProcessBuilder(*args)
        processBuilder.redirectErrorStream(true)
        val proc = processBuilder.start()
        Lok.debug("BashTools.stuffModifiedAfter.collecting.result")
        val result = ArrayList<AFile<*>>()
        val iterator = BashTools.inputStreamToFileIterator(proc.inputStream)
        while (iterator.hasNext()) {
            val path = iterator.next()
            Lok.debug(javaClass.simpleName + ".stuffModifiedAfter.collected: " + path)
            result.add(path)
        }
        Lok.debug("BashTools.stuffModifiedAfter.collecting.done")
        return result
    }

    @Throws(IOException::class)
    private fun exec(cmd: String): Iterator<AFile<*>> {
        val args = arrayOf(BIN_PATH, "-c", cmd)
        Lok.debug("BashToolsUnix.exec: $cmd")
        val proc = ProcessBuilder(*args).start()
        return BashTools.inputStreamToFileIterator(proc.inputStream)
    }

    @Throws(IOException::class)
    override fun find(directory: AFile<*>, pruneDir: AFile<*>): Iterator<AFile<*>> {
        return exec("find " + escapeQuotedAbsoluteFilePath(directory) + " -mindepth 1" + " -path " + escapeQuotedAbsoluteFilePath(pruneDir) + " -prune -o -print")
    }

    override fun stuffModifiedAfter(directory: AFile<*>, pruneDir: AFile<*>, timeStamp: Long): Iterator<AFile<*>>? {
        System.err.println("BashToolsUnix.stuffModifiedAfter()... I AM THE UNIX GUY! >:(")
        return null
    }

    @Throws(IOException::class)
    override fun mkdir(dir: AFile<*>) {
        val args = arrayOf(BIN_PATH, "-c", "mkdir " + escapeQuotedAbsoluteFilePath(dir))
        ProcessBuilder(*args).start()
    }

    @Throws(IOException::class)
    override fun mv(source: File, target: File): Boolean {
        val src = source.absolutePath.replace("'".toRegex(), "\\'")
        val tgt = target.absolutePath.replace("'".toRegex(), "\\'")
        val cmd = "mv '$src' '$tgt'"
        val args = arrayOf(BIN_PATH, "-c", cmd)
        val process = ProcessBuilder(*args).start()
        try {
            process.waitFor()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        val reader = BufferedReader(InputStreamReader(process.errorStream))
        val error = reader.readLine() ?: return true
        Lok.debug("BashToolsUnix.mv")
        return false
    }

    class ShareFolderProperties(val counted: Long, val containsSymLinks: Boolean)

    @Throws(IOException::class, InterruptedException::class)
    fun getShareFolderPropeties(dir: File): ShareFolderProperties {
        var count: Long = 0L
        var containsSymLinks = false;
        val thread = Thread(Runnable {
            dir.walkTopDown().onEnter { file ->
                if (Files.isSymbolicLink(Paths.get(file.toURI()))) {
                    containsSymLinks = true
                    false
                } else
                    file.isDirectory
            }.forEach { count++ }
        })
        thread.start()
        thread.join()
        return ShareFolderProperties(count, containsSymLinks)
    }

    companion object {

        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            AFile.configure(DefaultFileConfiguration())
            val f = AFile.instance("f")
            f.mkdirs()
            val bashToolsUnix = BashToolsUnix()
            val modifiedAndInode = bashToolsUnix.getFsBashDetails(f)
            Lok.debug("mod " + modifiedAndInode.modified + " ... inode " + modifiedAndInode.getiNode())
            f.delete()
        }
    }
}
